composer/core/state.py

```diff
@@ -540,14 +543,91 @@ def state_dict(self) -> Dict[str, Any]:
 
         return state_dict
 
-    def load_model_state(self, state_dict: Dict[str, Any], strict: bool):
+    def _apply_required_algorithms(
+        self,
+        state: Dict[str, Any],
+        logger: Logger,
+        exclude_algorithms: Optional[List[str]] = None,
+    ):
+        """Applies required algorithms which haven't been specified and aren't in the exclude list.
+
+        Args:
+            state (Dict[str, Any]): State from checkpoint.
+            logger (Logger): Logger to use.
+            exclude_algorithms (List[str], optional): List of algorithm names to exclude.
+        """
+        import composer.algorithms as algorithms  # type: ignore imports used in `eval(representation)`
+
+        # Get repr of existing algorithms
+        state_algos = {}
+        for algo in self.algorithms:
+            if algo.required_on_load():
+                if type(algo) not in state_algos:
+                    state_algos[type(algo)] = []
+                state_algos[type(algo)].append(algo.__repr__())
+
+        missing_algorithms = []
+        for algo_name, serialized_value in state['algorithms'].items():
+            try:
+                algo = eval(f"algorithms.{serialized_value['repr']}")
+                if algo.required_on_load() and (exclude_algorithms is None or
+                                                type(algo).__qualname__ not in exclude_algorithms):
+                    if type(algo) in state_algos and not serialized_value['repr'] in state_algos[type(algo)]:
+                        warnings.warn(
+                            textwrap.dedent(
+                                f"required_on_load algorithm {serialized_value['repr']} was enabled when training the "
+                                f"loaded checkpoint but is now specified in the following forms: {', '.join(state_algos[type(algo)])}."
+                                'Potential paarameter discrepencies for this required_on_load algorithm may lead to '
+                                'unexpected behavior, including failing to load weights for some layers.'))
+                    elif type(algo) not in state_algos:
+                        missing_algorithms.append((algo, serialized_value['repr']))
+            except AttributeError:
+                log.warning(
+                    f'Found unknown algorithm {algo_name}. Skipping check for if it is required when loading checkpoint.'
+                )
+
+        try:
+            for algo, algo_repr in missing_algorithms:
+                if algo.match(Event.INIT, self):
+                    algo.apply(Event.INIT, self, logger)
+                self.algorithms.append(algo)
```

Comment:
Do we need care about order here? How does engine interact/reorder now?

Reply:
@dskhudia I've added
1. reordering based on registered engine passes
2. error if this leads to an autoapply algorithm being sorted before an already applied algorithm